checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -176,6 +197,74 @@ public void addComputedTypeAnnotations(Tree tree, AnnotatedTypeMirror type, bool
         }
     }
 
+    @Override
+    public void preProcessClassTree(ClassTree classTree) {
+        super.preProcessClassTree(classTree);
+        Element elt = TreeUtils.elementFromDeclaration(classTree);
+        List<? extends Element> enclosedElts =
+                ElementUtils.enclosingClass(elt).getEnclosedElements();
+        List<FlowExpressions.Receiver> rs = new ArrayList<>();
+        for (Element e : enclosedElts) {
+            if (!ElementUtils.isEffectivelyFinal(e)) {
+                List<AnnotatedTypeMirror> enclosedTypes = findEnclosedTypes(e);
```

Comment:
Instead of recursively finding the enclosed types then parsing the expressions, I would do both at the same time.  Also, if you use a TreePathScanner, then you will have access to the TreePath, without have to recompute it via `getPath()`.  (Also, the path is still wrong for fields or methods of enclosed classes, because it is computed from the current element rather than the element on which the annotation appears.